clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/deploy/ops/job/RunDcosJobAtomicOperation.groovy

```diff
@@ -0,0 +1,164 @@
+package com.netflix.spinnaker.clouddriver.dcos.deploy.ops.job
+
+import com.netflix.spinnaker.clouddriver.data.task.Task
+import com.netflix.spinnaker.clouddriver.data.task.TaskRepository
+import com.netflix.spinnaker.clouddriver.dcos.DcosClientProvider
+import com.netflix.spinnaker.clouddriver.dcos.deploy.description.job.RunDcosJobDescription
+import com.netflix.spinnaker.clouddriver.deploy.DeploymentResult
+import com.netflix.spinnaker.clouddriver.orchestration.AtomicOperation
+import mesosphere.marathon.client.model.v2.LocalVolume
+import mesosphere.metronome.client.model.v1.Artifact
+import mesosphere.metronome.client.model.v1.Constraint
+import mesosphere.metronome.client.model.v1.Docker
+import mesosphere.metronome.client.model.v1.Job
+import mesosphere.metronome.client.model.v1.JobRunConfiguration
+import mesosphere.metronome.client.model.v1.JobSchedule
+import mesosphere.metronome.client.model.v1.Placement
+import mesosphere.metronome.client.model.v1.RestartPolicy
+
+class RunDcosJobAtomicOperation implements AtomicOperation<DeploymentResult> {
+  private static final String BASE_PHASE = "RUN_JOB"
+
+  final DcosClientProvider dcosClientProvider
+  final RunDcosJobDescription description
+
+  RunDcosJobAtomicOperation(DcosClientProvider dcosClientProvider, RunDcosJobDescription description) {
+    this.dcosClientProvider = dcosClientProvider
+    this.description = description
+  }
+
+  private static Task getTask() {
+    TaskRepository.threadLocalTask.get()
+  }
+
+  @Override
+  DeploymentResult operate(List priorOutputs) {
+    def jobName = description.general.id
+
+    task.updateStatus BASE_PHASE, "Initializing creation of job ${jobName}."
+
+    def dcosClient = dcosClientProvider.getDcosClient(description.credentials)
+
+    if (!dcosClient.maybeJob(description.general.id).isPresent()) {
+      task.updateStatus BASE_PHASE, "Job with id of ${jobName} does not exist, creating job."
+
+      def job = mapDescriptionToJob(description)
+
+      if (job.schedules) {
+        dcosClient.createJobWithSchedules(job)
+      } else {
+        dcosClient.createJob(job)
+      }
+
+      task.updateStatus BASE_PHASE, "Job ${jobName} was successfully created."
+    }
+
+    task.updateStatus BASE_PHASE, "Triggering job ${jobName}..."
+
+    def jobRun = dcosClient.triggerJobRun(jobName)
+
+    task.updateStatus BASE_PHASE, "Job ${jobName} has been started."
+
+    return new DeploymentResult().with {
+      deployedNames = [jobRun.id]
+      deployedNamesByLocation[jobRun.jobId] = [jobRun.id]
+      it
+    }
+  }
+
+  static Job mapDescriptionToJob(RunDcosJobDescription jobDescription) {
+    new Job().with {
+      id = jobDescription.general.id
+      it.description = jobDescription.general.description
+
+      if (jobDescription.labels) {
+        labels = jobDescription.labels.clone() as Map<String, String>
+      }
+
+      run = new JobRunConfiguration().with {
```

Comment:
So these are required, the reason I don't test for this in the validator is because to know whether we should be validating this is whether the job exists already or not. If the job is not already in DC/OS, this has to be there. Honestly I wasn't really sure how we should go about handling it at least with the validator.